keps/sig-apps/20200302-surge-on-daemonsets.md

```diff
@@ -0,0 +1,148 @@
+
+It should be possible for a node level integration to offer zero-downtime upgrades via a DaemonSet without resorting to a higher level orchestration.
+
+### Goals
+
+- Add support for Surge to the DaemonSet rolling update strategy
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+The design of Deployment rolling updates introduced the surge concept, and the initial design for DaemonSet updates considered the implications of adding the Surge strategy later (https://github.com/kubernetes/community/blob/master/contributors/design-proposals/apps/daemonset-update.md#future-plans).
+
+We would add `MaxSurge *intstr.IntOrString` to the RollingUpdate deployment. It would have a default value of 0, preserving current behavior. We would allow MaxUnavailable to be 0 when MaxSurge is set.
+
+```
+// Spec to control the desired behavior of daemon set rolling update.
+type RollingUpdateDaemonSet struct {
+	// The maximum number of DaemonSet pods that can be unavailable during the
+	// update. Value can be an absolute number (ex: 5) or a percentage of total
+	// number of DaemonSet pods at the start of the update (ex: 10%). Absolute
+	// number is calculated from percentage by rounding up.
+	// This cannot be 0 if MaxSurge is 0
+	// Default value is 1.
+	// Example: when this is set to 30%, at most 30% of the total number of nodes
+	// that should be running the daemon pod (i.e. status.desiredNumberScheduled)
+	// can have their pods stopped for an update at any given
+	// time. The update starts by stopping at most 30% of those DaemonSet pods
+	// and then brings up new DaemonSet pods in their place. Once the new pods
+	// are available, it then proceeds onto other DaemonSet pods, thus ensuring
+	// that at least 70% of original number of DaemonSet pods are available at
+	// all times during the update.
+	// +optional
+	MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty" protobuf:"bytes,1,opt,name=maxUnavailable"`
+
+	// The maximum number of nodes with an existing available DaemonSet pod that
+  // can have an updated DaemonSet pod during during an update.
+	// Value can be an absolute number (ex: 5) or a percentage of desired pods (ex: 10%).
+	// This can not be 0 if MaxUnavailable is 0.
+	// Absolute number is calculated from percentage by rounding up.
+	// Defaults to 25%.
+	// Example: when this is set to 30%, at most 30% of the total number of nodes
+	// that should be running the daemon pod (i.e. status.desiredNumberScheduled)
+	// can have their a new pod created before the old pod is marked as deleted.
+  // The update starts by launching new pods on 30% of nodes. Once an updated
+  // pod is available (Ready for at least minReadySeconds) the old DaemonSet pod
+  // on that node is marked deleted. If the old pod becomes unavailable for any
+  // reason (Ready transitions to false, is evicted, or is drained) an updated
+  // pod is immediatedly created on that node without considering surge limits.
```

Comment:
I see that this is following the availability reconciliation. An edge case is the resource consumption that could sky rocket on readiness blip. Like running 3 masters on one VM and this immediately doubles resources for intensive DS, although I've specified `MaxSurge: 1`.

Reply:
Only if you have an upgrade pending.  If you select surge != 0 you are allowing double resource consumption per node.  So you were going to double up resources on that node at some point in time `t` after `t0` (when you updated), all that the blip does is move it forward a bit faster.  If you have no upgrade pending, then we don't launch another pod.  

Agree the docs should call out resource consumption mindset.